exporters/pelorus/utils.py

```diff
@@ -0,0 +1,181 @@
+
+    You may specify the path as either an iterable of keys / indexes, or a single string.
+    The string will be split on '.' so you can emulate the nested attribute lookup `ResourceField`
+    would offer.
+
+    A `name` for the item, if specified, makes the error message in the exception more useful.
+
+    Kubernetes API items often are deeply nested, with any number of fields that could be absent.
+    When using an `openshift.dynamic.ResourceField`, it will turn attribute accesses into
+    dictionary accesses. Normally, a deeply nested access like item.status.ref.foo.bar has four different spots
+    you could get an `AttributeError`. With a `ResourceField`, there are actually only three, since `item.status`
+    will return `None` if `status` is absent, but `None` will not have a `ref` field, leading to an
+    AttributeError.
+
+    This all may be unnecessary once Python 3.11 comes out, because of PEP-0647:
+    https://www.python.org/dev/peps/pep-0657/
+    """
+    item = root
+    if isinstance(path, str):
+        # filter out leading dot (or accidental double dots, technically)
+        path = [part for part in path.split(".") if part]
+    for i, key in enumerate(path):
+        try:
+            item = item[key]
+        except (TypeError, IndexError, KeyError):
+            if default is not __GET_NESTED_NO_DEFAULT:
+                return default
+
+            raise BadAttributePathError(
+                root=root,
+                path=path,
+                path_slice=slice(i),
+                value=item,
+                root_name=name,
+            )
+
+    return item
+
+
+@dataclasses.dataclass
+class BadAttributePathError(Exception):
+    """
+    An error representing a nested lookup that went wrong.
+    """
+
+    root: Any
+    path: list[Any]
+    path_slice: slice
+    value: Any
+    root_name: Optional[str] = None
+
+    @property
+    def message(self):
+        return (
+            f"While trying to look up {'.'.join(self.path)} within "
+            f"{self.root_name if self.root_name else 'an item'}, "
+            f"{'.'.join(self.path[self.path_slice])} was {self.value}, "
+            "so we could not continue."
+        )
+
+
+class TypedString(UserString):
+    """
+    A "wrapped" string that can only test for equality between TypedStrings
+    of the same type, or a regular string.
+
+    This helps when strings representing different things are easily confused--
+    such as PipelineRun names versus Pipeline names.
+    """
+
+    def __eq__(self, other):
+        if isinstance(other, TypedString):
+            if type(self) is type(other):
+                return self.data == other.data
+            else:
+                return False
+        elif type(other) is str:
+            return self.data == other
+        else:
+            return NotImplemented
+
+    def __repr__(self):
```

Comment:
Should you define `__str__(self) `as well?

Reply:
I need to double check how these are used within the openshift library. I think I was unsure if `__str__` should transparently use the underlying string, or use `__repr__`.